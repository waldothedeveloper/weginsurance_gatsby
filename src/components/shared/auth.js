import React, { createContext, useContext, useEffect, useState } from "react";
import { getAuth, onAuthStateChanged, signOut } from "firebase/auth";

import { app } from "../../utils/firebaseConfig";
import { navigate } from "gatsby";

const auth = getAuth(app);

export const AuthContext = createContext();

export const AuthContextProvider = (props) => {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      setLoading(true);
      if (user) {
        setUser(user);
        setLoading(false);
      } else {
        setLoading(false);
      }
    });
    return () => unsubscribe();
  }, []);
  return <AuthContext.Provider value={{ user, loading }} {...props} />;
};

export const useAuthState = () => {
  const context = useContext(AuthContext);

  if (context === undefined) {
    throw new Error(`useAuthState must be used within an AuthContextProvider`);
  }
  return { ...context, isAuthenticated: context.user !== null };
};

export const logOut = () =>
  signOut(auth)
    .then(() => {
      navigate(`/auth/login`);
    })
    .catch((error) => {
      console.log(`error signing out: `, error);
    });
