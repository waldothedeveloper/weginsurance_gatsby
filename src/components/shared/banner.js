import {
  ExclamationCircleIcon,
  ExclamationIcon,
} from "@heroicons/react/outline";
import React, { useEffect, useState } from "react";

import { Fragment } from "react";
import PropTypes from "prop-types";
import { Transition } from "@headlessui/react";
import { XIcon } from "@heroicons/react/outline";
import { classNames } from "../../utils/classNames";

//
const notificationType = {
  error: `relative bg-red-500`,
  info: `relative bg-yellow-500`,
  success: `relative bg-sky-500`,
};
export const Banner = ({ type = null, message, dismissMessage }) => {
  const [isOpen, setIsOpen] = useState(false);

  useEffect(() => {
    if (message) {
      setIsOpen(true);
      // ! think about this
      // setTimeout(() => {
      //   setIsOpen(false);
      // }, 12000);
    }
  }, [message]);

  //
  return (
    type !== null && (
      <Transition
        data-testid="banner"
        show={isOpen}
        as={Fragment}
        enter="transform ease-out duration-300 transition"
        enterFrom="translate-y-2 opacity-0 sm:translate-y-0 sm:translate-x-2"
        enterTo="translate-y-0 opacity-100 sm:translate-x-0"
        leave="transition ease-in duration-100"
        leaveFrom="opacity-100"
        leaveTo="opacity-0"
      >
        <div className={notificationType[type]}>
          <div className="mx-auto max-w-7xl py-3 px-3 sm:px-6 lg:px-8">
            <div className="pr-16 sm:px-16 sm:text-center">
              <p className="flex items-center justify-center font-medium text-white">
                {type !== `info` ? (
                  <ExclamationCircleIcon
                    className="mr-1 h-6 w-6 text-red-50"
                    aria-hidden="true"
                  />
                ) : (
                  <ExclamationIcon
                    className="mr-1 h-6 w-6 text-yellow-50"
                    aria-hidden="true"
                  />
                )}
                <span className="inline">{message}</span>
              </p>
            </div>
            <div className="absolute inset-y-0 right-0 flex items-start pt-1 pr-1 sm:items-start sm:pt-1 sm:pr-2">
              <button
                data-testid="dismiss-button"
                onClick={() => {
                  dismissMessage(``);
                  setIsOpen(false);
                }}
                type="button"
                className={classNames(
                  type === `info`
                    ? `hover:bg-yellow-700`
                    : type === `error`
                    ? `hover:bg-red-700`
                    : `hover:bg-sky-700`,
                  `flex rounded-md p-2 focus:outline-none focus:ring-2 focus:ring-white`
                )}
              >
                <span className="sr-only">Dismiss</span>
                <XIcon className="h-6 w-6 text-white" aria-hidden="true" />
              </button>
            </div>
          </div>
        </div>
      </Transition>
    )
  );
};
Banner.propTypes = {
  type: PropTypes.oneOf([`error`, `info`, `success`, null]),
  message: PropTypes.string,
  dismissMessage: PropTypes.func,
};
