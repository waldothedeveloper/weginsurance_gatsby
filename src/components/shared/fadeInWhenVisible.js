import PropTypes from "prop-types";
import React from "react";
import { motion } from "framer-motion";
export const FadeInWhenVisible = ({ children }) => (
  <motion.div
    initial="hidden"
    whileInView="visible"
    viewport={{ once: true }}
    transition={{ duration: 1.2 }}
    variants={{
      visible: { opacity: 1 },
      hidden: { opacity: 0 },
    }}
  >
    {children}
  </motion.div>
);

FadeInWhenVisible.propTypes = {
  children: PropTypes.node.isRequired,
};
