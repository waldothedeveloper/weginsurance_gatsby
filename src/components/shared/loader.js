import React from "react";

export const Loader = () => (
  <div className="flex min-h-screen items-center justify-center">
    <div className="bg-white">
      <div className="mx-auto max-w-7xl py-16 px-4 sm:py-24 sm:px-6 lg:px-8">
        <div className="text-center">
          <h2 className="animate-pulse text-base font-bold uppercase tracking-wide text-sky-600">
            Cargando...
          </h2>
          <p className="mt-1 text-4xl font-extrabold text-sky-900 sm:tracking-tight">
            Un momento por favor
          </p>
          <svg
            className="mx-auto mt-5 h-8 w-8 max-w-xl animate-spin text-sky-600"
            xmlns="http://www.w3.org/2000/svg"
            fill="none"
            viewBox="0 0 24 24"
          >
            <circle
              className="opacity-25"
              cx="12"
              cy="12"
              r="10"
              stroke="currentColor"
              strokeWidth="4"
            ></circle>
            <path
              className="opacity-75"
              fill="currentColor"
              d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
            ></path>
          </svg>
        </div>
      </div>
    </div>
  </div>
);
