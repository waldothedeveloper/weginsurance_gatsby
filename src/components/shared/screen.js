import React from "react";
import propTypes from "prop-types";
export const SharedScreen = ({
  rightPanel: RightPanel,
  leftPanel: LeftPanel,
}) => (
  <div className="z-0 mx-auto grid grid-cols-12 gap-8 px-6">
    <div className="col-span-4 h-full rounded-lg border-2 border-solid border-gray-100 py-6 lg:px-4">
      {LeftPanel}
    </div>
    <div className="col-span-8 h-full rounded-lg border-2 border-solid border-gray-100 pt-6 lg:px-1">
      {RightPanel}
    </div>
  </div>
);

SharedScreen.propTypes = {
  rightPanel: propTypes.node.isRequired,
  leftPanel: propTypes.node.isRequired,
};
