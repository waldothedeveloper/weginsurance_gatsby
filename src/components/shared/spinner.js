import PropTypes from "prop-types";
import React from "react";
export const Spinner = ({ children }) => (
  <div className="flex flex-col items-center justify-center">
    <svg
      className="-ml-1 mr-3 h-8 w-8 animate-spin text-slate-400"
      xmlns="http://www.w3.org/2000/svg"
      fill="none"
      viewBox="0 0 24 24"
    >
      <circle
        className="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        strokeWidth="4"
      />
      <path
        className="opacity-75"
        fill="currentColor"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
      />
    </svg>
    <div className="mt-4 font-medium text-slate-800">{children}</div>
  </div>
);

Spinner.propTypes = {
  children: PropTypes.node,
};
