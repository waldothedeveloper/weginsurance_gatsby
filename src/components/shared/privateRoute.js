import { Loader } from "./loader";
import PropTypes from "prop-types";
import React from "react";
import { isBrowser } from "../../utils/firebaseConfig";
import { navigate } from "gatsby";
import { useAuthState } from "../shared/auth";

const PrivateRoute = ({ component: Component, ...props }) => {
  const { loading, isAuthenticated } = useAuthState();

  if (isBrowser && !isAuthenticated && !loading) {
    navigate(`/auth/login`);
  }

  if (loading) {
    return <Loader />;
  } else {
    return <Component {...props} />;
  }
};

export default PrivateRoute;
PrivateRoute.propTypes = {
  component: PropTypes.any,
};
