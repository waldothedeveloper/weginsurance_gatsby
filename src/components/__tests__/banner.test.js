import { render, screen } from "@testing-library/react";

import { Banner } from "../shared/banner";
import React from "react";
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/react";

//
const user = userEvent.setup();

//
describe(`Banner`, () => {
  it(`should render correctly`, () => {
    const { container } = render(
      <Banner
        type="error"
        message="Error message"
        dismissMessage={() => false}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it(`should render information message if type = info`, () => {
    render(
      <Banner
        type="info"
        message="Error message"
        dismissMessage={() => false}
      />
    );

    expect(screen.getByText(/error/i)).toBeInTheDocument();
  });

  it(`should not render if type is null or undefined`, () => {
    render(<Banner type={undefined} message="" dismissMessage={() => false} />);

    expect(screen.queryByTestId(`banner`)).not.toBeInTheDocument();
  });

  it(`should close when dismiss button is clicked`, async () => {
    const dismissMessage = jest.fn();
    render(
      <Banner
        type="error"
        message="Error message"
        dismissMessage={dismissMessage}
      />
    );

    const dismissButton = screen.queryByTestId(`dismiss-button`);
    await user.click(dismissButton);

    await waitFor(() => {
      expect(dismissMessage).toHaveBeenCalled();
    });

    await waitFor(() => {
      expect(screen.queryByTestId(`banner`)).not.toBeInTheDocument();
    });
  });
});
