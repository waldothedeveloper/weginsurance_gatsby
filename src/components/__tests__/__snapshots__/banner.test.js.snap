// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner should render correctly 1`] = `
<div>
  <div
    class="relative bg-red-500 transform ease-out duration-300 transition translate-y-2 opacity-0 sm:translate-y-0 sm:translate-x-2"
    data-testid="banner"
  >
    <div
      class="mx-auto max-w-7xl py-3 px-3 sm:px-6 lg:px-8"
    >
      <div
        class="pr-16 sm:px-16 sm:text-center"
      >
        <p
          class="flex items-center justify-center font-medium text-white"
        >
          <svg
            aria-hidden="true"
            class="mr-1 h-6 w-6 text-red-50"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span
            class="inline"
          >
            Error message
          </span>
        </p>
      </div>
      <div
        class="absolute inset-y-0 right-0 flex items-start pt-1 pr-1 sm:items-start sm:pt-1 sm:pr-2"
      >
        <button
          class="hover:bg-red-700 flex rounded-md p-2 focus:outline-none focus:ring-2 focus:ring-white"
          data-testid="dismiss-button"
          type="button"
        >
          <span
            class="sr-only"
          >
            Dismiss
          </span>
          <svg
            aria-hidden="true"
            class="h-6 w-6 text-white"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
